tests/utils/log_reader.py

```diff
@@ -27,77 +28,109 @@
-            while not self._stop_event.is_set():
-                _line = self._file.readline()
-                if _line:
-                    return _line
-                time.sleep(0.1)
-
-        get_line()
-        with self._start_cv:
-            self._has_data = True
-            self._start_cv.notify()
-        self._file.seek(0)
-        while not self._stop_event.is_set():
-            line = get_line()
-            self.handle_new_line(line)
-
-    def start(self, wait_for_data=False):
-        super(LogReader, self).start()
-        if not wait_for_data:
-            return
-        with self._start_cv:
-            while not self._has_data:
-                self._start_cv.wait(0.1)
-        return
+        self._error_line_patterns = dict()
+        self._lines = list()
+
+    def _check_line_for_error(self, line):
+        for pattern, (compiled_pattern, message) in self._error_line_patterns.items():
+            if compiled_pattern.match(line):
+                if not message:
+                    message = "Log file {0} contains error line: '{1}'. Pattern: {2}.".format(
+                        self._file_path, line, pattern
+                    )
+                raise LogReaderError(message)
+
+    def _handle_new_line(self, line):
+        self._lines.append(line)
+        self._check_line_for_error(line)
+
+    def _read_line(self):
+        line = self._file.readline()
+        if line:
+            self._handle_new_line(line)
+
+        return line
+
+    def wait_for_next_line(self, timeout=10):
+        timeout_time = time.time() + timeout
+        while True:
+            line = self._read_line()
+            if line:
+                return line
+
+            if time.time() >= timeout_time:
+                raise LogReaderTimeoutError(
+                    "Timeout of %s seconds reached while waiting for metrics" % timeout
+                )
+
+            time.sleep(0.01)
+
+    def go_to_end(self):
+        while True:
+            line = self._read_line()
+            if not line:
+                break
+
+        pass
+
+    def wait_for_lines(self, line_number):
+        for _ in range(line_number):
+            self.wait_for_next_line()
 
     @property
-    def lines(self):
-        with self._lines_lock:
-            return list(self._lines)
+    def last_line(self):
+        if len(self._lines):
+            return self._lines[-1]
+
+    def wait_for_matching_line(self, pattern, timeout=10):
+        compiled_pattern = re.compile(pattern)
+        while True:
+            line = self.wait_for_next_line(timeout=timeout)
+            if compiled_pattern.match(line):
+                return line
+
+    def wait(self, seconds):
+        """
+        Keep checking for new lines for some period of time in 'seconds'.
+        """
+        stop_time = time.time() + seconds
+        while True:
+            try:
+                self.wait_for_next_line()
+            except LogReaderTimeoutError:
```

Comment:
Should we log a message here, perhaps under debug?

Reply:
This code part was refactored :)